### PR TITLE
Make poll() with POLLOUT return immediately

### LINKED ISSUE #18867: zsock_poll() unnecessarily wait when querying for ZSOCK_POLLOUT
**Describe the bug**
When `zsock_poll()` is called to wait for the `ZSOCK_POLLOUT` event, and the socket is already ready for sending, it does not return immediately, but waits for the entirety of the `timeout` passed to it.

An analogous bug also affects `zsock_select()`.

This has been tested on the QEMU emulated target, and on an ST Nucleo-F429ZI board that is equipped with an Ethernet port. I have not tested it on non-Ethernet interfaces (e.g. WiFi, cellular, etc.).

**To Reproduce**
1. Build the following application:

**CMakeLists.txt**:
```cmake
cmake_minimum_required(VERSION 3.13.1)
include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
project(pollout_bug_demo)
target_sources(app PRIVATE main.c)
```

**prj.conf**:
```
CONFIG_NETWORKING=y
CONFIG_NET_IPV4=y
CONFIG_NET_SOCKETS=y
CONFIG_NET_CONFIG_NEED_IPV4=y
CONFIG_NET_CONFIG_SETTINGS=y
CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1"
```

**main.c**:
```c
#include <assert.h>
#include <net/socket.h>

void main(void) {
    struct sockaddr_in addr = {
        .sin_family = AF_INET,
        .sin_port = htons(1234)
    };
    int ret = zsock_inet_pton(AF_INET, "192.0.2.2", &addr.sin_addr);
    assert(!ret);

    int fd = zsock_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
    assert(fd >= 0);
    ret = zsock_connect(fd, (const struct sockaddr *) &addr, sizeof(addr));
    assert(!ret);
    
    printf("Connected!\n");
    
    struct zsock_pollfd pollfd = {
        .fd = fd,
        .events = ZSOCK_POLLOUT
    };
    ret = zsock_poll(&pollfd, 1, 5000);
    assert(ret == 1);
    assert(pollfd.revents == ZSOCK_POLLOUT);
    
    zsock_send(fd, "Hello, world", 12, 0);

    printf("Packet sent!\n");
}
```

2. Run it on an Ethernet-equipped board, or the QEMU emulator. On the 192.0.2.2 host, run something like `nc -u -l 1234`
3. Observe the 5-second delay between "Connected!" and "Packet sent!" messages. If you change the timeout passed to `zsock_poll()`, the delay will change accordingly.

An analogous application using `zsock_select()` instead of `zsock_poll()` can be written, exhibiting the same behaviour. Also, changing the socket type from UDP to TCP also shows the same problem.

**Expected behavior**
`zsock_poll()` shall return immediately, as equivalent calls do on Linux, macOS and Windows. In this usage, it shall block only if the internal socket buffers are indeed full.

**Impact**
Some network-related libraries that would be otherwise trivial to port to Zephyr cause the code to wait for extremely long if `poll()` is used to wait for the `POLLOUT` event (or an equivalent `select()` call).

**Environment (please complete the following information):**
 - OS: Ubuntu 18.04
 - Toolchain: Zephyr SDK 0.10.3
 - Commit SHA or Version used: `aa9b762d12`

### PR DESCRIPTION
**Pull request by jukkar**
Fixes #18867

### CODE REVIEW

**Comment by pfalcon** [comment]
I would be very careful with changes like this. It's sad if we had a bug like reported, but careless "fixing" it may make the overall situation only worse. Mind that "return immediately" may also mean "try send(), have it fall immediately, and call poll() again immediately". Which will return immediately. And all that will be infinite loop, or at least busy-loop with 100% CPU burn.

I'd suggest that time has come to fix it properly, and predicate POLLOUT handling on the actual availability of the output buffers.

All in all, I'd like to spend extra time reviewing both current code and proposed change. (Please don't be surprised if I -1 it to allow for more time to do that if there will be premature approvals.)

**Comment by jukkar** [comment]
For UDP socket, we can always return immediately. For TCP we should check if we can really send, but it is not easy thing to do here. It will not change the outcome at all from current behaviour as we do not check if we can send atm after a timeout. We can add more checks for TCP in following PRs.

**Comment by pfalcon** [comment]
> For UDP socket, we can always return immediately.

poll() requires returning when you can actually perform a requested action, so "we can always return immediately" is not true.

> It will not change the outcome at all from current behaviour

Please read above - you're turning into slow latent bug into hot, burn-your-fingers-by-touching-cpu bug. That's not exactly "not changing behavior".

> We can add more checks 

We can definitely turn the code into sieve of workarounds, but I suggest we start to solve well-known, well-cooked technical debt.

And beyond that (if there's a pledge "let's not solve bugs! let's work around them! please! please!"), I'd need to check that the code you have here doesn't introduce the outright infinite loop.

**Comment by jukkar** [comment]
> hot, burn-your-fingers-by-touching-cpu bug

Where did you get this? If the application busy loops then one needs to fix it of course.

**Comment by pfalcon** [comment]
> Where did you get this? If the application busy loops then one needs to fix it of course.

I explained how a typical poll-based event loop works. If there's a case: a) that app doesn't process an even returned by poll; or b) poll returns a false event, then there will be a loop of calls to poll() with no progress made on each iteration, which is a definition of infinite loop. If there're also no delays in there, then it will be 100% CPU use busy-loop.

**Comment by jukkar** [comment]
> poll() requires returning when you can actually perform a requested action, so "we can always return immediately" is not true.

Yes, I will refine the checks a bit. In practice we can write to the socket in these scenarios:
* UDP: if the socket is open
* TCP: if the socket is open and there is room in the window to write data
* any other protocol type: if the socket is open (this could be refined if there is need)

**Comment by pfalcon** [comment]
> In practice we can write to the socket in these scenarios:

First and foremost, we can write to a socket if there's an available output buffer. And that's the missing check, and what I call as "time to add". To remind, this is a known problem (https://github.com/zephyrproject-rtos/zephyr/issues/5842), the proper solution of which requires adding a per-socket "send buffer", so we always can guarantee that a socket can send stuff independently of others (i.e. not being DoSed by other sockets/connections), and that we can check properly whether it's writable (or if previous data is still being sent).

I (of course) didn't go for that solution during initial socket implementation, because the aim was to implement sockets as a very thin layer on top of native API with minimal additional resource usage, and poll() support seemed like distant advanced feature to care about. I still think that we aren't yet ready for per-socket buffers (I'd assume https://github.com/zephyrproject-rtos/zephyr/pull/11443 would take care of that). But IMHO it's time to resolve #5842 nonetheless, because changes like this may only make situation worse.

So, look for # of free buffers in system output buffer pool, and return POLLOUT only if they're >0 (yes, return POLLOUT for every input POLLOUT, even if there's only 1 buffer available - sockets will compete for it, but that's ok, as long as on each iteration, at least one will get it (and actually use); again, the proper solution is outlined above).

**Comment by jukkar** [comment]
> First and foremost, we can write to a socket if there's an available output buffer. And that's the missing check, and what I call as "time to add". To remind, this is a known problem (#5842), the proper solution of which requires adding a per-socket "send buffer", so we always can guarantee that a socket can send stuff independently of others (i.e. not being DoSed by other sockets/connections), and that we can check properly whether it's writable (or if previous data is still being sent).

I doubt we will ever have socket specific buffers, it would be too wasteful for the global resources.

> 
> So, look for # of free buffers in system output buffer pool, and return POLLOUT only if they're >0 (yes, return POLLOUT for every input POLLOUT, even if there's only 1 buffer available - sockets will compete for it, but that's ok, as long as on each iteration, at least one will get it (and actually use); again, the proper solution is outlined above).

Nak, that will not work. Buffers come and go and if we run out of buffers, the application would not be send/recv anything anyway.

I think what I am proposing is a good compromise.

**Comment by pfalcon** [comment]
> I doubt we will ever have socket specific buffers, it would be too wasteful for the global resources.

That's required for reliable networking, so we'll have them sooner or later, e.g. when enough no. of CVEs will be issued against us.

> Buffers come and go and if we run out of buffers, the application would not be send/recv anything anyway.

Application would not be send/recv, it would go into infinite loop, as described multiple times, and in detail, above. So any heuristic helping to avoid otherwise imminent infinite loop is helpful.

> I think what I am proposing is a good compromise.

I'm not sure what you're referring to exactly, but this PR *exposes* the latent issue with possible 100% cpu-burning infinite loop (as was said a few time already too).

File: subsys/net/lib/sockets/sockets.c [hunk 1]
```diff
@@ -915,6 +918,28 @@ static int zsock_poll_prepare_ctx(struct net_context *ctx,
...
+		}
+
+#if defined(CONFIG_NET_TCP)
+		if (net_context_get_ip_proto(ctx) == IPPROTO_TCP &&
+		    ctx->tcp && net_tcp_get_recv_wnd(ctx->tcp) == 0) {
+			;
+		} else
+#endif
+		{
+			errno = EINPROGRESS;
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: What's the meaning of this? You cannot return arbitrary errors from prepare_ctx() vmethod, such an error will be fatal for poll(), and fatal errors for poll() should be minimized (in particular, problem with a *single* socket should never lead to abort of entire poll())

**Review by pfalcon — CHANGES_REQUESTED** [review]
Found at least one issue, looking for more.

**Comment by pfalcon** [comment]
@jukkar, Please confirm that you read this comment: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/sockets/sockets.c#L990 , and fully aware/understand of it, and intend to make this and any future changes to poll() with it mind. Failing that means unconditional infinite loop. That's my biggest concern with this change.

File: subsys/net/lib/sockets/sockets.c [hunk 2]
```diff
@@ -999,6 +1024,15 @@ int z_impl_zsock_poll(struct zsock_pollfd *fds, int nfds, int timeout)
 				continue;
 			}
 
+			/* The EINPROGRESS is returned if POLLOUT is done
+			 * to indicate that we can write to the socket
+			 * immediately.
+			 */
+			if (errno == EINPROGRESS) {
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: Lol, so you didn't look at few lines immediately above, and just flopped a new code doing absolutely the same. Golden! So you see why I'm skeptical-by-default about such changes? Nobody gives a damn about existing code. Reading code is for cowards. Patch, patch, patch, patch!

**Review by pfalcon — DISMISSED** [review]

File: tests/net/socket/poll/prj.conf [hunk 3]
```diff
@@ -21,3 +21,6 @@ CONFIG_MAIN_STACK_SIZE=2048
 CONFIG_ZTEST=y
 
 CONFIG_QEMU_TICKLESS_WORKAROUND=y
+
```
Comment (pfalcon) [inline tests/net/socket/poll/prj.conf]: What exact problem you have without these options? This tests is supposed (and does) run fine without these options with standard Zephyr qemu networking setup. And relying on such setup makes it easier to hack on this test (and adds a bit more of coverage under different setup than CI).

File: subsys/net/lib/sockets/sockets.c (hunk 1 above)
Comment (jukkar) [reply subsys/net/lib/sockets/sockets.c]: We can certainly return EALREADY here too, the patch just returns EINPROGRESS just to differentiate POLLIN and POLLOUT.

File: subsys/net/lib/sockets/sockets.c (hunk 2 above)
Comment (jukkar) [reply subsys/net/lib/sockets/sockets.c]: > Lol, so you didn't look at few lines immediately above, and just flopped a new code doing absolutely the same. Golden! So you see why I'm skeptical-by-default about such changes? Nobody gives a damn about existing code. Reading code is for cowards. Patch, patch, patch, patch!

I would appreciate that you stop making these kind of comments. These are not professional and productive.

File: tests/net/socket/poll/prj.conf (hunk 3 above)
Comment (jukkar) [reply tests/net/socket/poll/prj.conf]: If the settings are in testcase.yaml file, then you cannot run the test directly from command line.

File: subsys/net/lib/sockets/sockets.c (hunk 2 above)
Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: Do you hint that writing such code is professional and productive?

File: tests/net/socket/poll/prj.conf (hunk 3 above)
Comment (pfalcon) [reply tests/net/socket/poll/prj.conf]: No, you can run *this test* from command line. Just try, and see.

File: subsys/net/lib/sockets/sockets.c (hunk 1 above)
Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: > We can certainly return EALREADY here too, the patch just returns EINPROGRESS just to differentiate POLLIN and POLLOUT.

There's no need to introduce additional notions. EALREADY is the API contract for signaling behavior of "Ignore the incoming timeout param, and perform underlying k_poll() call with K_NO_WAIT, but be sure to still perform". If this fact is not clear enough and you see where it can be documented better (perhaps, in a code comments, as this is internal API contract), then +1 for doing that.

File: tests/net/socket/poll/prj.conf (hunk 3 above)
Comment (jukkar) [reply tests/net/socket/poll/prj.conf]: Been there, done that. Sure it works if you setup slip etc. Running the test standalone requires CONFIG_NET_TEST=y and CONFIG_NET_LOOPBACK=y so it makes sense to place them to prj.conf so that they are always available.

Comment (pfalcon) [reply tests/net/socket/poll/prj.conf]: > Sure it works if you setup slip etc.

Yes. And if usually run slip setup anyway, then it should be pretty seamless. I thus would suggest to avoid making changes unless really needed - again, having it both ways allows for better coverage. That's just a suggestion though. (But other things in this PR are truly -1's.)

Comment (jukkar) [reply tests/net/socket/poll/prj.conf]: > And if usually run slip setup anyway, then it should be pretty seamless. I thus would suggest to avoid making changes unless really needed - again, having it both ways allows for better coverage.

That is just your usecase, for me I usually do not use slip any longer. So NAK to this comment.

**Comment by jukkar** [comment]
Lets revisit this again now that other TCP related fixes have been merged.

**Comment by pfalcon** [comment]
Yes, we definitely need to proceed with this.

1. But I would think that the only initial change which would be required is:

~~~
	if (pfd->events & ZSOCK_POLLOUT) {
			errno = EALREADY;
			ret = -1;
}
~~~

That would fix the original bug report (but leave concerns of possible busy-looping which I shared earlier).

2. If you'd like to handle a case of empty TCP **send** window (i.e. a peer says that it can't receive anything so far), then unfortunately we can't do that now, due to https://github.com/zephyrproject-rtos/zephyr/issues/3441 . Using **our** receive window for this check (code quoted below) doesn't look correct to me:

~~~
#if defined(CONFIG_NET_TCP)
		if (net_context_get_ip_proto(ctx) == IPPROTO_TCP &&
		    ctx->tcp && net_tcp_get_recv_wnd(ctx->tcp) == 0) {
			;
		} else
#endif
~~~

Even if it would be, please remember that poll() handling is split in 2 parts: the "prepare" IOCTL call, but then also "update" IOCTL. Updating "prepare" part without paired update to "update" part may lead to incorrect functioning, and it looks like that case to me here.

3. To continue on that, the code below:

~~~
		(*pev)->mode = K_POLL_MODE_NOTIFY_ONLY;
		(*pev)->type = K_POLL_TYPE_IGNORE;
		(*pev)++;
	}
~~~

Looks both superfluous and problematic due to the issue explained in p.2 ("prepare" handler adds new stuff to k_poll array, which "update" handler isn't ready to handle (as it's not updated in this PR)).


Summing up: I'd suggest to restart work on this part from scratch, with the code quoted in p.1 above, and carefully consider if anything beyond it is needed/makes sense. (We definitely can make further conclusions from this case, e.g. to say that poll() handling isn't documented (commented) clearly enough. I'd appreciate such feedback and any suggestions that it should be commented better, or feel free to include it in this patch).

**Comment by pfalcon** [comment]
We just discussed this with @jukkar during Zephyr F2F in Lyon, and there is an agreement re: points shared above. The new version with fixes is expected. As I'm on vacation next week, removing my -1 to not block this going into 2.1.

**Comment by jukkar** [comment]
>     1. To continue on that, the code below:
> 
> 
> ```
> 		(*pev)->mode = K_POLL_MODE_NOTIFY_ONLY;
> 		(*pev)->type = K_POLL_TYPE_IGNORE;
> 		(*pev)++;
> 	}
> ```
> 
> Looks both superfluous and problematic due to the issue explained in p.2 ("prepare" handler adds new stuff to k_poll array, which "update" handler isn't ready to handle (as it's not updated in this PR)).

Originally I did not had that code, but added it because of this crash
```
ASSERTION FAIL [num_events > 0] @ ZEPHYR_BASE/kernel/poll.c:257
	zero events
```
(just remembered this one after I removed these code lines and run the tests).
We need to figure out how to solve this.

**Comment by jukkar** [comment]
Revisited the PR after discussion with Paul. I did not remove the setting of event struct as that caused crashes from kernel.

**Comment by pfalcon** [comment]
> ASSERTION FAIL [num_events > 0] @ ZEPHYR_BASE/kernel/poll.c:257
	zero events

I see. Well, it's quite an artificial assert, a case of zero input `struct k_poll_event` can appear naturally when filtering some application array, like our case here shows. And we apparently should resolve it by adding POSIX behavior, i.e. k_poll() with 0 event structures should be equivalent to k_sleep() for the requested timeout. @jukkar, can you make such a change and include it here (IMHO that's a case when such a fixup can go together with existing PR and doesn't have to be separate. YMMV).

**Comment by pfalcon** [comment]
> And we apparently should resolve it by adding POSIX behavior, i.e. k_poll() with 0 event structures should be equivalent to k_sleep()

Of course, we can add a check around a call to k_poll(), but IMHO it's better to fix it once on k_poll() level.

**Comment by jukkar** [comment]
Changing poll.c like this seems to work ok
```
diff --git a/kernel/poll.c b/kernel/poll.c
index 608887a3ac..0b1cffbcfc 100644
--- a/kernel/poll.c
+++ b/kernel/poll.c
@@ -254,7 +254,7 @@ int z_impl_k_poll(struct k_poll_event *events, int num_events, s32_t timeout)
 
        __ASSERT(!z_arch_is_in_isr(), "");
        __ASSERT(events != NULL, "NULL events\n");
-       __ASSERT(num_events > 0, "zero events\n");
+       __ASSERT(num_events >= 0, "<0 events\n");
 
        events_registered = register_events(events, num_events, &poller,
                                            (timeout == K_NO_WAIT));
```
I will do some additional testing.

**Comment by pfalcon** [comment]
~~~
-       __ASSERT(num_events > 0, "zero events\n");
+       __ASSERT(num_events >= 0, "<0 events\n");
~~~

Hopefully that'll work. From quick look, I don't see anything obvious in that func or immediate callees which wouldn't work with 0 passed-in structure, but kernel folks would need to review.

**Comment by jukkar** [comment]
New version with fixes to k_poll()

**Comment by jukkar** [comment]
ping @andrewboie / @andyross are you ok with the assert change in k_poll()?

**Review by pfalcon — APPROVED** [review]

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
I would like this patch a lot better if tests/kernel/poll was updated to cover this scenario.

**Review by andyross — APPROVED** [review]
The changes makes perfect sense to me.  @andrewboie is correct that this is a data coverage issue, as we never actually call this with a num_events of zero.  But it's actually not technically a code coverage problem as the only use of that value is as the bounds of a for() loop which is executed identically (i.e. with zero iterations).  Personally I think it's fine, but throwing in a "unconditional sleep using k_poll()" test case certainly wouldn't hurt.

**Comment by jukkar** [comment]
Added k_poll() tests.

**Comment by jukkar** [comment]
ping @andrewboie, can you re-check this one

**Comment by jukkar** [comment]
ping @andrewboie

**Review by tyrantlizard77 — APPROVED** [review]
